### DISCUSSION TITLE
MPY: can´t not flash

### DISCUSSION THREAD

**Discussion by tcpipchip** [micropython/micropython#11708] [STM32 / Pyboard] [answered]
![image](https://github.com/micropython/micropython/assets/11488460/9f64eacf-8f18-4537-a614-390cf6fb8cb4)

How can i enable on STM32WL55 ?

**Comment by tcpipchip** [comment]
>>> help('modules')
__main__          micropython       ubinascii         uplatform
_uasyncio         pyb               ucollections      urandom
builtins          uarray            uctypes           ure
gc                uasyncio/__init__ uerrno            uselect
lora/__init__     uasyncio/core     uhashlib          ustruct
lora/async_modem  uasyncio/event    uio               usys
lora/modem        uasyncio/funcs    ujson             utime
lora/stm32wl5     uasyncio/lock     umachine          uzlib
lora/sx126x       uasyncio/stream   uos
Plus any modules on the filesystem
>>> 
>>>

**Reply by davefes** [reply]
Maybe:

`>>> import lora`
`>>> dir(lora)`

A bit confused about what you are trying to do ... is not the message `Can't mount flash` something that needs sorting first?

**Comment by tcpipchip** [comment]
sorry, is about the **MPY: can´t moun flash**

This making not get success to transfer a script to execute on FS

![image](https://github.com/micropython/micropython/assets/11488460/be641327-cbd5-4bac-b74e-f1b8e00aa303)

Does need a directive during STM32 build  to make a automaticly format on FLASH and make a FS ?

**Comment by davefes** [comment]
Just curious ... have you tried this image?
https://micropython.org/download/NUCLEO_WL55/

**Comment by tcpipchip** [comment]
yes, same FLASH problem
![image](https://github.com/micropython/micropython/assets/11488460/ade3fd71-346d-4c30-8a4f-dcda2e00e4cf)
and still dont have LoRa

**Reply by davefes** [reply]
Sorry, I have no experience with the Pyboard and thought that it must have a filesystem like the ESP32.
https://docs.micropython.org/en/latest/pyboard/general.html

> If there is no SD card, then it uses the internal filesystem /flash as the boot filesystem, otherwise, it uses the SD card /sd. After the boot, the current directory is set to one of the directories above.

Maybe, looking around the old forum:
https://forum.micropython.org/viewforum.php?f=6
will be more helpful.

Good luck!

**Comment by tcpipchip** [comment]
hi
I am using the STM32WL55, not the pyboard, but is the same family processor...i agree...now i am investigating if the 
MICROPY_VFS_FAT = 0 must to set to 1

**Reply by davefes** [reply]
I thought that most or all boards used LittleFS for the filesystem.

**Comment by tcpipchip** [comment]
Yes! 8 k should be enough

**Reply by davefes** [reply]
Don't know where 8K comes from my ESP32 boards have 2MB.

How far do you get to in this tutorial before problems surface?
https://docs.micropython.org/en/latest/pyboard/tutorial/script.html

Also, there might be some differences between a Pyboard's hardware and your hardware that is not permitting mounting the flash.  I read about SD card maybe having some effect on the way the filesystem mounts.

**Reply by davefes** [reply]
BTW, have you got a link to that LoRa distribution?  I looked through all his repositories and could not find anything LoRa.

Thanks

**Reply by tcpipchip** [reply]
https://lom204-cli-wisol.blogspot.com/2023/06/lsm110a-sendo-programado-em-micropython.html

References

**Reply by davefes** [reply]
Thank you.

**Comment by projectgus** [comment]
> MPY: Can't mount flash

The STM32WL55 is a pretty low-resource device, so it doesn't have a flash filesystem automatically. You don't always need a filesystem in the internal flash, for example if you freeze all of the Python code into the firmware binary and flash it that way.

There is support in the default WL55 config for Littlefs2, so you can format a Littlefs image in the available flash space like this:

```py
import os, pyb
os.VfsLfs2.mkfs(pyb.Flash(start=0))
```

... then when you next restart (for example, with Ctrl-D)  this filesystem will be mounted at `/flash` and no more errors.  The [documentation for this feature is here](https://docs.micropython.org/en/latest/reference/filesystem.html#littlefs).

Note the available flash will be small, on my system it was 24KB. Run `pyb.Flash(start=0)` to print the value of the Flash object, which includes the length.

---

Regarding testing the lora functionality from [the lora pull requests](https://github.com/micropython/micropython-lib/pull/572), you already have lora frozen into the firmware so you are mostly done. You don't even need a filesystem to start testing it (I didn't create any filesystem on my NUCLEO-WL55 board until just now.) 

To run an example: 

First edit `./lib/micropython-lib/micropython/lora/examples/simple_rxtx/simple_rxtx.py` (or another example) to use the correct modem for your board. Inside the file, find the `get_modem()` function and replace it like this:

```py
def get_modem():
    from lora import WL55SubGhzModem

    # The LoRa configuration will depend on your board and location, see
    # README under "Modem Configuration" for some possible examples.
    lora_cfg = { 'freq_khz': SEE_README_FOR_CORRECT_VALUE }
    return WL55SubGhzModem(lora_cfg)    
```

and then run

```sh
cd ./lib/micropython-lib/micropython/lora/examples/simple_rxtx/
mpremote run simple_rxtx.py
```

.. to run the script over the mpremote link.

To make the LoRa code persistent, you can either use the filesystem or freeze your Python code into the firmware build. See the Micropython docs for more about this.

Until the PRs merge, the README for the lora driver (with stm32wl55 support) is [here](https://github.com/projectgus/micropython-lib/tree/feature/lora_stm32wl5/micropython/lora#readme).

**Reply by tcpipchip** [reply]
Thank you so much @projectgus 
Thanks again!
I hope to do tests soon and finish the blog. Btw, format works :)

btw, plans to LoraWAN ?

**Reply by projectgus** [reply, marked as answer]
Great to hear. Please mark my comment as a answer if it helped (you'll find a button under my comment for this.)

> plans to LoraWAN ?

I would love to see LoraWAN on Micropython, but for now I want to get the basic LoRa support merged. Then we'll see.

**Comment by tcpipchip** [comment]
I am getting error
![image](https://github.com/micropython/micropython/assets/11488460/568268d0-54b8-44a7-b94a-352a6edac8ef)

**Comment by tcpipchip** [comment]
i executed help(lora)
![image](https://github.com/micropython/micropython/assets/11488460/7691ed2e-e2e6-4726-b4d9-4db6a78044d5)

**Comment by tcpipchip** [comment]
i had to modify to

# MicroPython lora simple_rxtx example - asynchronous API version
```
# MIT license; Copyright (c) 2023 Angus Gratton
import uasyncio
from umachine import Pin, SPI

import micropython

micropython.alloc_emergency_exception_buf(256)

def get_async_modem():
    from lora import AsyncWL55SubGhzModem
    # The LoRa configuration will depend on your board and location, see
    # README under "Modem Configuration" for some possible examples.
    lora_cfg = { 'freq_khz': 916000 }
    return AsyncWL55SubGhzModem(lora_cfg)    

async def main_task():
    modem = get_async_modem()
    await uasyncio.gather(
        uasyncio.create_task(send_coro(modem)),
        uasyncio.create_task(recv_coro(modem)),
    )

async def recv_coro(modem):
    while True:
        print("Receiving...")
        rx = await modem.recv(2000)
        if rx:
            print(f"Received: {repr(rx)}")
        else:
            print("Receive timeout!")


async def send_coro(modem):
    counter = 0
    while True:
        print("Sending...")
        await modem.send(f"Hello world from async MicroPython #{counter}".encode())
        print("Sent!")
        await uasyncio.sleep(5)
        counter += 1
        
if __name__ == "__main__":
    uasyncio.run(main_task()) 
```

**Comment by tcpipchip** [comment]
![image](https://github.com/micropython/micropython/assets/11488460/e21b7bc2-5dad-4e54-8f9a-bfc77edad988)

**Comment by tcpipchip** [comment]
Tomorrow I will test the communication

**Comment by tcpipchip** [comment]
testing...no luck here on STMWL55JC (LSM110A)
![image](https://github.com/micropython/micropython/assets/11488460/6e9590c1-5960-4734-bd88-31ccf029b630)
![image](https://github.com/micropython/micropython/assets/11488460/22016ddb-893b-4e93-936c-2e9be638d564)

**Comment by tcpipchip** [comment]
Hi @projectgus 

I add stm32lorawan arduino example to receive the package from micropython
![image](https://github.com/micropython/micropython/assets/11488460/3bbe4c06-5c95-4248-9799-eb62a35bef00)

**Reply by tcpipchip** [reply]
still investigating...

**Comment by tcpipchip** [comment]
is that ok ?
I see time out
![image](https://github.com/micropython/micropython/assets/11488460/b42f8054-862d-461d-b6ac-dc87b8bf5387)

**Comment by tcpipchip** [comment]
time out with 6000 value,  syncronized

![image](https://github.com/micropython/micropython/assets/11488460/5cbf9f03-7515-4fd0-9611-f8107e2d9d02)

**Comment by tcpipchip** [comment]
one of my problems was the the power supply...
Now all looks working good

**Comment by tcpipchip** [comment]
![image](https://github.com/micropython/micropython/assets/11488460/3ffc7c87-4bfd-45ac-95b1-98bd012ce9b8)

**Comment by tcpipchip** [comment]
![image](https://github.com/micropython/micropython/assets/11488460/c8b640ec-591e-43dc-b68d-ba8c7a049f22)

**Comment by tcpipchip** [comment]
Angus
This is off topic
but did you already the error
Corrupted dir pair at {0x0, 0x1} on STM32WL55 ? (using the LITTLEFS) ?
It´s a problem that we are having o MESHTASTIC
Thanks

**Reply by projectgus** [reply]
No, I've not really used littlefs onSTM32WL55 @tcpipchip . Suggest opening another question about it, with as many details as you can provide. I feel like it might be because the total size for littlefs on this board is so small.

**Comment by tcpipchip** [comment]
how are going the LoRaWAN tests with STM32W55 ?

**Reply by projectgus** [reply]
No work on LoRaWAN for micropython-lib at the moment, I'm afraid. Would still like to support this, but it's not a priority.

**Comment by tcpipchip** [comment]
No problems! Just curiuos
